source/geod24/concurrency.d

```diff
@@ -1012,56 +1178,154 @@ private:
+                    auto cur_timeout = cur_fiber.blocker.getTimeout();
+
+                    // Keep track of the earliest timeout in the system
+                    if (cur_timeout != MonoTime.init
+                            && (earliest_timeout == MonoTime.init || cur_timeout < earliest_timeout))
+                    {
+                        earliest_timeout = cur_timeout;
+                    }
+
+                    this.wait_list.insert(cur_fiber);
                     continue;
                 }
-                else
+                else if (t)
+                {
+                    // We are exiting the dispatch loop prematurely, all resources
+                    // held by Fibers should be released.
+                    this.releaseResources(cur_fiber);
+                    throw t;
+                }
+
+                if (cur_fiber.state != Fiber.State.TERM)
                 {
-                    m_fibers[pos].sem = null;
+                    this.readyq.insert(cur_fiber);
                 }
+
+                // See if there are Fibers to be woken up if we reach a timeout
+                // or the scheduler semaphore was notified
+                if (MonoTime.currTime >= earliest_timeout || this.sem.tryWait())
+                    earliest_timeout = wakeFibers();
             }
 
-            auto t = m_fibers[pos].call(Fiber.Rethrow.no);
-            if (t !is null)
+            if (!this.wait_list.empty())
             {
-                throw t;
+                Duration time_to_timeout = earliest_timeout - MonoTime.currTime;
+
+                // Sleep until a timeout or an event
+                if (earliest_timeout == MonoTime.init)
+                    this.sem.wait();
+                else if (time_to_timeout > 0.seconds)
+                    this.sem.wait(time_to_timeout);
             }
-            if (m_fibers[pos].state == Fiber.State.TERM)
+
+            // OS Thread woke up populate ready queue
+            earliest_timeout = wakeFibers();
+        }
+    }
+
+    /***********************************************************************
+
+        Move unblocked Fibers to ready queue
+
+        Return:
+            Returns the earliest timeout left in the waiting list
+
+    ***********************************************************************/
+
+    MonoTime wakeFibers()
+    {
+        import std.range;
+        MonoTime earliest_timeout;
+
+        auto wait_range = this.wait_list[];
+        while (!wait_range.empty)
+        {
+            auto fiber = wait_range.front;
+
+            // Remove the unblocked Fiber from wait list and
+            // append it to the end ofready queue
+            if (!fiber.blocker.shouldBlock())
             {
-                if (pos >= (m_fibers = remove(m_fibers, pos)).length)
-                    pos = 0;
+                this.wait_list.popFirstOf(wait_range);
+                fiber.blocker = null;
+                this.readyq.insert(fiber);
```

Comment:
What about popping from the `wait_range`? It's not needed here?

Reply:
`popFirstOf` does that too. It removes both from the list and the range.